traversal/procedure/GraphProcedure.java

```diff
@@ -214,74 +150,160 @@ public boolean equals(Object o) {
+                }
+            });
+            plannerVertex.ins().forEach(plannerEdge -> {
+                if (!registeredEdges.contains(plannerEdge) && plannerEdge.isSelected()) {
+                    registeredEdges.add(plannerEdge);
+                    adjacents.add(plannerEdge.from());
+                    registerEdge(plannerEdge);
+                }
+            });
+            plannerVertex.loops().forEach(plannerEdge -> {
+                if (!registeredEdges.contains(plannerEdge) && plannerEdge.isSelected()) {
+                    registeredEdges.add(plannerEdge);
+                    registerEdge(plannerEdge);
+                }
+            });
+            adjacents.forEach(v -> registerVertex(v, registeredVertices, registeredEdges));
+        }
+
+        private void registerEdge(PlannerEdge.Directional<?, ?> plannerEdge) {
+            ProcedureVertex<?, ?> from = vertex(plannerEdge.from());
+            ProcedureVertex<?, ?> to = vertex(plannerEdge.to());
+            ProcedureEdge<?, ?> edge = ProcedureEdge.of(from, to, plannerEdge);
+            registerEdge(edge);
+        }
+
+        public void registerEdge(ProcedureEdge<?, ?> edge) {
+            if (edge.from().equals(edge.to())) {
+                edge.from().loop(edge);
+            } else {
+                edge.from().out(edge);
+                edge.to().in(edge);
+            }
+        }
+
+        private ProcedureVertex<?, ?> vertex(PlannerVertex<?> plannerVertex) {
+            if (plannerVertex.isThing()) return thingVertex(plannerVertex.asThing());
+            else return typeVertex(plannerVertex.asType());
+        }
+
+        private ProcedureVertex.Thing thingVertex(PlannerVertex.Thing plannerVertex) {
+            return thingVertex(plannerVertex.id());
         }
 
-        public ProcedureVertex.Type labelledType(String label) {
-            return labelledType(label, false);
+        private ProcedureVertex.Type typeVertex(PlannerVertex.Type plannerVertex) {
+            return typeVertex(plannerVertex.id());
         }
 
-        public ProcedureVertex.Type labelledType(String label, boolean isStart) {
-            return typeVertex(Identifier.Variable.of(Reference.label(label)), isStart);
+        private ProcedureVertex.Thing thingVertex(Identifier identifier) {
+            return vertices.computeIfAbsent(identifier, ProcedureVertex.Thing::new).asThing();
         }
 
-        public ProcedureVertex.Type namedType(String name) {
-            return namedType(name, false);
+        private ProcedureVertex.Type typeVertex(Identifier identifier) {
+            return vertices.computeIfAbsent(identifier, ProcedureVertex.Type::new).asType();
         }
 
-        public ProcedureVertex.Type namedType(String name, boolean isStart) {
-            return typeVertex(Identifier.Variable.of(Reference.name(name)), isStart);
+        // ---- manual builder methods ----
+
+        public ProcedureVertex.Type labelledType(int order, String label) {
```

Comment:
note: a separate set of builder methods exist for manually building procedures. Here the vertices directly get the order number.